### ISSUE TITLE
Implement ontology-aware evaluation metrics for ENVO predictions (partial credit for parent/child terms)

### ISSUE THREAD

**Issue by turbomam** [contextualizer-ai/env-embeddings#51]
## Problem

Current RF evaluation uses binary exact match scoring:
- Predicted term matches true term → 1.0 (correct)
- Predicted term differs from true term → 0.0 (incorrect)

This penalizes **semantically close predictions** equally with **completely wrong predictions**.

### Example

```
True: ENVO:00000447 (marine biome)

Predictions:
  - ENVO:00000447 (marine biome) → Score: 1.0 ✅ exact match
  - ENVO:00000428 (biome) → Score: 0.0 ❌ [parent term, very close!]
  - ENVO:01000048 (coastal sea water) → Score: 0.0 ❌ [child term, close!]
  - ENVO:01000174 (terrestrial biome) → Score: 0.0 ❌ [sibling, somewhat related]
  - ENVO:00002011 (cropland soil) → Score: 0.0 ❌ [unrelated]
```

The parent term and child term predictions should get **partial credit** - they're much closer than completely unrelated terms.

## Proposed Solution

Implement **hierarchy-aware scoring** using ENVO ontology structure.

### 1. Ontology Distance Scoring

Score based on shortest path in ENVO graph:

```python
def hierarchical_score(predicted, true, ontology):
    """
    Score based on ontology relationship.
    
    Returns:
        1.00 - Exact match
        0.75 - Parent or child (1 hop)
        0.50 - Grandparent, grandchild, or sibling (2 hops)
        0.25 - Distant relative (3+ hops)
        0.00 - Unrelated (no path)
    """
```

### 2. Semantic Similarity Scoring

Use existing ENVO embeddings (1536-dim) for similarity:

```python
def semantic_similarity_score(predicted, true, envo_embeddings):
    """
    Score based on cosine similarity between ENVO term embeddings.
    
    Returns:
        0.9 - Very similar (cosine > 0.9)
        0.6 - Moderately similar (cosine > 0.7)
        0.0 - Not similar (cosine < 0.7)
    """
```

### 3. New Evaluation Metrics

Add to RF notebook evaluation:

- **hierarchical_accuracy**: Mean of hierarchy-aware scores
- **semantic_accuracy**: Mean of embedding-similarity scores
- **top_k_hierarchical**: % where true term in top-k by ontology distance
- **Confusion analysis**: Which classes have "close misses" vs "completely wrong"

## Implementation Tasks

- [ ] Add `pronto` dependency for ENVO.owl parsing
- [ ] Create `src/env_embeddings/ontology_metrics.py` module
  - [ ] `ENVOHierarchy` class with graph building
  - [ ] `hierarchical_score()` function
  - [ ] `semantic_similarity_score()` function
  - [ ] `hierarchical_accuracy()` metric
- [ ] Download/cache ENVO.owl in data directory
- [ ] Update RF notebook (random_forest_envo_prediction.ipynb):
  - [ ] Add ontology-aware scoring section
  - [ ] Compare exact vs hierarchical accuracy
  - [ ] Visualize: heatmap with hierarchy coloring
  - [ ] Show examples of "close misses" that get partial credit
- [ ] Document methodology in notebook markdown cells
- [ ] Add tests for ontology metrics

## Use Cases

1. **More nuanced RF evaluation**: Distinguish "close" from "completely wrong"
2. **Active learning**: Focus manual review on truly confused samples, not near-misses
3. **Better k-NN comparison**: k-NN may predict related terms via majority vote
4. **Inform model improvements**: If many "close" predictions, model understands context but needs refinement

## Expected Results

For each of 3 scales (broad, local, medium):
- Exact match accuracy: ~50-60% (baseline from previous runs)
- Hierarchical accuracy: ~65-75% (with partial credit)
- Semantic accuracy: ~60-70% (embedding-based)

**Insight**: If hierarchical >> exact, model is learning environmental context but struggling with fine distinctions.

## Dependencies

```toml
# Add to pyproject.toml
pronto = "^2.5.0"  # ENVO OWL parsing
```

## Related

- #49 - RF training on NMDC (needs this for evaluation)
- #31 - Original classifier proposal
- Existing ENVO integration in `envo_embeddings.py`

## Priority

**High** - Improves evaluation quality immediately, no retraining needed, high scientific value.

**Comment by justaddcoffee** [comment]
I was going to suggest this one too:

```
2. Semantic Similarity Scoring
Use existing ENVO embeddings (1536-dim) for similarity:

def semantic_similarity_score(predicted, true, envo_embeddings):
    """
    Score based on cosine similarity between ENVO term embeddings.
    
    Returns:
        0.9 - Very similar (cosine > 0.9)
        0.6 - Moderately similar (cosine > 0.7)
        0.0 - Not similar (cosine < 0.7)
    """
```

But I think we should use this `semantic_similarity_score` as the loss function during training (as opposed to measuring this after the fact to measure how well the RF does)

**Comment by turbomam** [comment]
I thought about this more today. I don't think we should reward near misses just because they are close in OLS embeddings space

If a submitter annotates a Biosample with `env_medium` of 'dry soil', but the RF predicts a processual class like 'digging a hole' because that contains a description like 

> A planned process in which a shovel is thrust into the earth and soil is discarded into a pile

That's not nearly as good as is it guessed the parent class 'soil' or the child class 'extremely dry soil'

GPT5 says that scoring RF performance on hierarchical classification tasks like that is pretty well established, so I won't paste it all here.

**Comment by justaddcoffee** [comment]
From the ELDER work, I suspect that embedding space would be pretty good at not conflating different uses of words like "soil", because the descriptions and the embeddings of the descriptions are pretty good at capturing nuances like this

But, I don't think we need to go wild with making a great RF classifier with my fancy new loss function

Maybe we can just sync up with Emiley / Shreyas / Patrick and try to deploy our RF (and also our other deterministic methods for things like retrieving temperature and elevation) to help with NMDC curation

**Comment by turbomam** [comment]
A hop count of 2 (especially over `is_a` *and* `part_of`) between C1 and C2 may not mean the same thing as a hop count of 2 between C3 to C4.

-> Switch to MRCA information content

**Comment by turbomam** [comment]
-log(freq) of every node

**Comment by turbomam** [comment]
After validation train on 100% and then deploy

**Comment by justaddcoffee** [comment]
My $0.02: I am dubious of whether we can use hops like this, since hops are not a reliable metric of semantic distance (1 hop in a certain part of the ontology might be a huge semantic leap, but might be a small semantic leap in a different part), but I'd be glad to be wrong. I think using -log(frequency) of common ancestor is the "right" way to give partial credit to the RF

But, our RF doesn't need to be perfect here, so your hop method or just the simple method we've used thus far (score = 1 for exact match, score = 0 otherwise) might be okay

**Comment by turbomam** [comment]
## Information Content (IC) Based Metrics - Analysis

After implementing the hop-based hierarchical scoring, I can confirm it's working correctly and providing improvements over exact-match scoring. However, the concern about non-uniform semantic distance per hop is valid.

### Why Hop-Based Scoring Has Limitations

The fundamental issue: 1 hop can represent vastly different amounts of semantic information loss depending on where you are in the ontology:
- Hopping from "marine sediment" → "sediment" loses significant specificity
- Hopping from "terrestrial biome" → "biome" loses less information (both are already general)

### Information Content (IC) Based Approaches

IC-based metrics address this by weighting the ontology structure with actual information content. The key insight: **a term's informativeness is inversely related to how common/general it is**.

**Two common IC calculation methods:**

1. **Corpus-based IC**: `IC(term) = -log(P(term))`
   - P(term) = frequency of term in actual data (GOLD/NMDC/NCBI datasets)
   - Rare, specific terms get high IC
   - Common, general terms get low IC

2. **Structure-based IC**: `IC(term) = -log((descendants(term) + 1) / total_terms)`
   - Based purely on ontology structure
   - Terms with many descendants are more general (low IC)

**Popular IC-based similarity metrics:**

- **Resnik**: `sim(t1, t2) = IC(LCA(t1, t2))`
  - Simplest: just the IC of the lowest common ancestor
  
- **Wu-Palmer**: `sim(t1, t2) = 2 × IC(LCA) / (IC(t1) + IC(t2))`
  - Normalized version considering depth of both terms
  
- **Lin**: `sim(t1, t2) = 2 × IC(LCA) / (IC(t1) + IC(t2))`
  - Similar to Wu-Palmer with different normalization

### Advantages for ENVO Evaluation

1. **Data-driven**: Using corpus-based IC, we can weight by actual term usage in our datasets
2. **Handles ontology asymmetry**: A specific→general mistake is penalized more than general→general
3. **Well-established**: These metrics are standard in bioinformatics (GO term similarity, etc.)
4. **Interpretable**: IC values have clear meaning (bits of information)

### Potential Hybrid Approach

We could combine approaches:
- Use hop distance for unrelated terms (no partial credit if too far)
- Use IC-based similarity for nearby terms (nuanced scoring for close predictions)
- Fall back to embedding similarity for cross-branch comparisons

Would be interested in implementing IC-based metrics to compare against the current hop-based approach. We have the data (term frequencies) and the tools (oaklib for ancestor traversal).

---
*Comment generated by Claude Code during debugging session*
